python/lsst/daf/butler/core/datasets/ref.py

```diff
@@ -0,0 +1,278 @@
+        # the Query infrastructure has a hard time obtaining Run objects, but
+        # that will change.
+        self.id = id
+        self.datasetType = datasetType
+        if conform:
+            self.dataId = DataCoordinate.standardize(dataId, graph=datasetType.dimensions)
+        else:
+            self.dataId = dataId
+        self._components = dict()
+        if components is not None:
+            self._components.update(components)
+        self.run = run
+        if hash is not None:
+            self._hash = hash
+        return self
+
+    def __eq__(self, other: DatasetRef):
+        return (self.datasetType, self.dataId, self.id) == (other.datasetType, other.dataId, other.id)
+
+    def __hash__(self) -> int:
+        return hash(self.datasetType, self.dataId, self.id)
+
+    def __repr__(self):
+        return f"DatasetRef({self.datasetType}, {self.dataId}, id={self.id}, run={self.run})"
+
+    @property
+    def hash(self) -> bytes:
+        """Secure hash of the `DatasetType` name and data ID (`bytes`).
+        """
+        if not hasattr(self, "_hash"):
+            message = hashlib.blake2b(digest_size=32)
+            message.update(self.datasetType.name.encode("utf8"))
+            self.dataId.fingerprint(message.update)
+            self._hash = message.digest()
+        return self._hash
+
+    @property
+    def components(self) -> Mapping[str, DatasetRef]:
+        """Named `DatasetRef` components.
+
+        Read-only (but not immutable); update via `Registry.attachComponent()`.
+        """
+        return MappingProxyType(self._components)
+
+    @property
+    def dimensions(self) -> DimensionGraph:
+        """The dimensions associated with the underlying `DatasetType`
+        """
+        return self.datasetType.dimensions
+
+    def __str__(self) -> str:
+        components = ""
+        if self.components:
+            components = ", components=[" + ", ".join(self.components) + "]"
+        return "DatasetRef({}, id={}, dataId={} {})".format(self.datasetType.name,
+                                                            self.id, self.dataId, components)
+
+    def __getnewargs_ex__(self) -> Tuple[Tuple[Any, ...], Dict[str, Any]]:
+        return ((self.datasetType, self.dataId),
+                {"id": self.id, "run": self.run, "components": self._components})
+
+    def resolved(self, id: int, run: Run, components: Optional[Mapping[str, DatasetRef]] = None
+                 ) -> DatasetRef:
+        """Return a new `DatasetRef` with the same data ID and dataset type
+        and the given ID and run.
+
+        Parameters
+        ----------
+        id : `int`
+            The unique integer identifier assigned when the dataset is created.
+        run : `Run`
+            The run this dataset was associated with when it was created.
+        components : `dict`, optional
+            A dictionary mapping component name to a `DatasetRef` for that
```

Comment:
I think this should also mention that the components supplied here overwrite any existing components but also merged with the existing ones. If this is a resolved datasetRef should we check that all the components have resolved dataset refs? Should we check they have the same run? What if there are some unresolved components and this call only overwrites some of them with resolved versions?

Reply:
So, I'm pretty sure we have concrete uses for:
 - resolved `DatasetRefs` with components
 - resolved `DatasetRefs` without components
 - unresolved `DatasetRefs` without components

where by "without components" I mean that the `DatasetRef` doesn't know about any components it might have, even if they exist, because it's just a wasteful to do anything with them because we know they won't be used.

The dataset-insertion interplay between `Registry` and `Datastore` is also a lot simpler if we allow the components to be updated `DatasetRef` (and I think that's okay w.r.t. immutability, since components aren't used in `__eq__`/`__hash__`).

Anyhow, with all that in mind, how about this:
 - We set `components` to `None` on an unresolved `DatasetRef`.  We raise in `__init__` if components are passed without `id` and `run`, and merge in `resolved()` only if re-resolving an already resolved `DatasetRef`.
 - We check that all component `DatasetRefs` are resolved, and that names are valid according to the storage class.

Reply:
Sounds good to me.

Reply:
Sounds good to me.